Userland/Libraries/LibCompress/Brotli.cpp

```diff
@@ -0,0 +1,905 @@
+/*
+ * Copyright (c) 2022, Michiel Visser <opensource@webmichiel.nl>
+ *
+ * SPDX-License-Identifier: BSD-2-Clause
+ */
+
+#include <AK/BinarySearch.h>
+#include <AK/QuickSort.h>
+#include <LibCompress/Brotli.h>
+#include <LibCompress/BrotliDictionary.h>
+
+namespace Compress {
+
+ErrorOr<size_t> BrotliDecompressionStream::CanonicalCode::read_symbol(LittleEndianInputBitStream& input_stream)
+{
+    size_t code_bits = 1;
+
+    while (code_bits < (1 << 16)) {
+        // FIXME: This is very inefficient and could greatly be improved by implementing this
+        //        algorithm: https://www.hanshq.net/zip.html#huffdec
+        size_t index;
+        if (binary_search(m_symbol_codes.span(), code_bits, &index))
+            return m_symbol_values[index];
```

Comment:
This index can be outside the vector, I get this when going to `cloudflare.com` sometimes, but do not have an example file on hand.
```
12.509 RequestServer(39:39): ASSERTION FAILED: i < m_size
../.././AK/Vector.h:149
12.509 [#0 RequestServer(39:39)]: Terminating RequestServer(39) due to signal 6
12.517 [#0 FinalizerTask(4:4)]: Backtrace:
0x00000000deadc0de  Kernel::Processor::switch_context(Kernel::Thread*&, Kernel::Thread*&) + 0x383
12.517 [#0 FinalizerTask(4:4)]: Generating coredump for pid: 39
12.552 CrashDaemon(9:9): New coredump file: /tmp/coredump/RequestServer_39_1649194427
12.659 WebContent(42:42): Parse error! [handle_initial @ ../../Userland/Libraries/LibWeb/HTML/Parser/HTMLParser.cpp:469]
12.713 CrashReporter(41:41): Started thread "", tid = 43
13.687 CrashReporter(41:43): Generating backtrace took 969 ms
13.687 CrashReporter(41:43): --- Backtrace for thread #0 (TID 39) ---
13.687 CrashReporter(41:43): 0x000000028d35a02b: [libsystem.so] syscall2 +0xb (syscall.cpp:25 => syscall.cpp:24)
13.687 CrashReporter(41:43): 0x0000000b0999eba7: [libc.so] abort +0x27 (stdlib.cpp:223)
13.687 CrashReporter(41:43): 0x0000000b099a3b3e: [libc.so] __assertion_failed +0x7e (assert.cpp:33)
13.690 CrashReporter(41:43): 0x000000166c084117: [libcompress.so.serenity] Compress::BrotliDecompressionStream::CanonicalCode::read_symbol(Core::Stream::LittleEndianInputBitStream&) [clone .localalias] +0x197 (Vector.h:149 => Vector.h:157 => Brotli.cpp:23)
13.690 CrashReporter(41:43): 0x000000166c086904: [libcompress.so.serenity] Compress::BrotliDecompressionStream::block_update_length(Compress::BrotliDecompressionStream::Block&) +0x264 (Brotli.cpp:473)
13.690 CrashReporter(41:43): 0x000000166c086c52: [libcompress.so.serenity] Compress::BrotliDecompressionStream::read_block_configuration(Compress::BrotliDecompressionStream::Block&) +0xf2 (Brotli.cpp:462)
13.690 CrashReporter(41:43): 0x000000166c087d3c: [libcompress.so.serenity] Compress::BrotliDecompressionStream::read(AK::Span<unsigned char>) +0x107c (Brotli.cpp:647)
13.690 CrashReporter(41:43): 0x0000000f66613289: [libcore.so.serenity] Core::Stream::Stream::read_all() +0x69 (Stream.cpp:57)
13.695 CrashReporter(41:43): 0x000000109b68a0f4: [libhttp.so.serenity] HTTP::Job::finish_up() [clone .localalias] +0xff4 (Job.cpp:86 => Job.cpp:613)
13.695 CrashReporter(41:43): 0x000000109b68ce4f: [libhttp.so.serenity] HTTP::Job::on_socket_connected()::{lambda()#2}::operator()() const +0x251f (Job.cpp:323)
13.695 CrashReporter(41:43): 0x000000109b6875c3: [libhttp.so.serenity] AK::Function<void ()>::CallableWrapper<HTTP::Job::register_on_ready_to_read(AK::Function<void ()>)::{lambda()#1}>::call() +0x63 (Function.h:91)
13.695 CrashReporter(41:43): 0x000000109b687880: [libhttp.so.serenity] AK::Function<void ()>::CallableWrapper<HTTP::Job::register_on_ready_to_read(AK::Function<void ()>)::{lambda()#1}::operator()() const::{lambda()#2}>::call() +0x50 (Function.h:91)
13.699 CrashReporter(41:43): 0x0000000f6660880c: [libcore.so.serenity] AK::Function<void ()>::CallableWrapper<Core::Object::deferred_invoke(AK::Function<void ()>)::{lambda()#1}>::call() +0x23c (Function.h:91)
13.699 CrashReporter(41:43): 0x0000000f665f1a6a: [libcore.so.serenity] Core::EventLoop::pump(Core::EventLoop::WaitMode) [clone .localalias] +0x28a (Function.h:91)
13.699 CrashReporter(41:43): 0x0000000f665f2229: [libcore.so.serenity] Core::EventLoop::exec() +0x119 (EventLoop.cpp:438)
13.699 CrashReporter(41:43): 0x0000001a965a05c1: [/bin/RequestServer] serenity_main(Main::Arguments) +0x1b1 (main.cpp:51)
13.699 CrashReporter(41:43): 0x0000001a96573915: [/bin/RequestServer] main +0x95 (Main.cpp:39)
13.702 CrashReporter(41:43): 0x0000001a96573b5f: [/bin/RequestServer] _entry +0x6f (crt0.cpp:49)
```